libs/ml/src/clustering/tsne.cpp

```diff
@@ -0,0 +1,497 @@
+typename TSNE<TensorType>::DataType TSNE<TensorType>::GetRandom(DataType /*mean*/,
+                                                                DataType /*standard_deviation*/)
+{
+  // TODO(issue 752): use normal distribution random instead
+  return rng_.AsType<DataType>();
+}
+
+/**
+ * i.e. Calculates the gradient of the Kullback-Leibler divergence between P and the Student-t
+ * based joint probability distribution Q
+ * @param output_matrix output low dimensional values tensor
+ * @param input_symmetric_affinities Tensor of symmetrized input matrix pairwise affinities
+ * @param output_symmetric_affinities Tensor of symmetrized output matrix pairwise affinities
+ * @param num Tensor of precalculated values num[i,j]=1+||yi-yj||^2
+ * @param ret return value output_matrix shaped tensor of gradient values.
+ */
+template <typename TensorType>
+TensorType TSNE<TensorType>::ComputeGradient(TensorType const &output_matrix,
+                                             TensorType const &input_symmetric_affinities,
+                                             TensorType const &output_symmetric_affinities,
+                                             TensorType const &num)
+{
+  assert(input_matrix_.shape().at(0) == output_matrix.shape().at(0));
+
+  TensorType ret(output_matrix.shape());
+
+  for (SizeType i{0}; i < output_matrix.shape().at(0); i++)
+  {
+    TensorType output(output_matrix.shape().at(1));
+    for (SizeType j{0}; j < output_matrix.shape().at(0); j++)
+    {
+      if (i == j)
+      {
+        continue;
+      }
+
+      DataType p_i_j;
+      p_i_j = input_symmetric_affinities.At(i, j);
+
+      DataType q_i_j = output_symmetric_affinities.At(i, j);
+
+      // (Pij-Qij)
+      DataType val = p_i_j - q_i_j;
+
+      // /(1+||yi-yj||^2)
+      fetch::math::Multiply(num.At(i, j), val, val);
+
+      // val*(yi-yj), where val=(Pij-Qij)/(1+||yi-yj||^2)
+
+      TensorType diff = (output_matrix.Slice(j).Copy()) - (output_matrix.Slice(i).Copy());
+
+      SizeVector shape = diff.shape();
+      shape.erase(shape.begin());
+      diff.Reshape(shape);
+
+      output += Multiply(val, diff);
+    }
+
+    for (SizeType k = 0; k < output_matrix.shape().at(1); k++)
+    {
+      ret(i, k) = fetch::math::Multiply(static_cast<DataType>(-1), output.At(k));
+    }
+  }
+
+  return ret;
+}
+
+/**
+ * i.e. if any value of input matrix is lower than min, sets it to min
+ * @param matrix input tensor
+ * @param min limit value to be applied
+ * @return
+ */
+template <typename TensorType>
+void TSNE<TensorType>::LimitMin(TensorType &matrix, DataType const &min)
+{
+  for (auto &val : matrix)
+  {
+    val = (val < min) ? min : val;
+  }
+}
+
+///////////////////////////////
+/// EXPLICIT INSTANTIATIONS ///
+///////////////////////////////
+
+template class TSNE<math::Tensor<int8_t>>;
+template class TSNE<math::Tensor<int16_t>>;
+template class TSNE<math::Tensor<int32_t>>;
+template class TSNE<math::Tensor<int64_t>>;
```

Comment:
what about unsigned types?

Reply:
initially I ignored these because the code has a DataType{-2} in it; but I found a way to re-arrange the math to not require that, so I can include unsigned ints. Huzzah!